src/overlays/actors/ovl_Ball/ac_ball.c

```diff
@@ -22,63 +55,752 @@ ActorProfile Ball_Profile = {
+                    this->collider.attribute.dim.unk_2 = 0x14;
+                    this->collider.attribute.dim.radius = 0x12;
+                    this->actor.colStatus.mass = 0xFE;
+                    this->actor.speed = 0.0f;
+                    return;
+                }
+            }
+        }
+    } else {
+        mRlib_Get_norm_Clif(&this->actor, &bgNorm);
+        add_calc0(&this->height, 0.5f, 100.0f);
+    }
+
+    if (!(fabsf(bgNorm.x) < 0.008f) || !(fabsf(bgNorm.z) < 0.008f)) {
+        if (Math3d_normalizeXyz_t(&bgNorm)) {
+            this->actor.velocity.x += 1.35f * bgNorm.x;
+            this->actor.velocity.z += 1.35f * bgNorm.z;
+            mRlib_spdXZ_to_spdF_Angle(&this->actor.velocity, &this->maxSpeed, &this->actor.world.rot.y);
+            if (this->maxSpeed < 8.0f) {
+                this->maxSpeed = this->maxSpeed;
+            } else {
+                this->maxSpeed = 8.0f;
+            }
+            this->acceleration = 0.1f;
+        }
+    } else {
+        this->maxSpeed = 0.0f;
+        this->acceleration = 0.12f;
+    }
+
+    this->actor.terminalVelocity = -20.0f;
+    this->actor.gravity = 0.6f;
+    this->speed = this->actor.speed;
+    if (this->actor.colResult.inWater || this->actor.colResult.unk5 == 0xB) {
+        if (this->actor.colResult.onGround) {
+            aBALL_process_ground_water_init(this, game_play);
+        } else {
+            this->bounceTimer = 0;
+            aBALL_process_air_water_init(this, game_play);
+        }
+    } else if (!this->actor.colResult.onGround) {
+        this->bounceTimer = 0;
+        aBALL_process_air_init(this, game_play);
+        return;
+    }
+
+    if (!(game_play->state.frameCounter & 7) && this->actor.colResult.unk5 == 9) {
+        if (this->actor.speed > 1.0f) {
+
+            if (this->actor.speed > 4.0f) {
+                var_v0 = 1;
+            } else {
+                var_v0 = 0;
+            }
+
+            common_data.clip.unk_090->unk_00(0x33, this->actor.world.pos, 1, this->actor.world.rot.y, game_play,
+                                             this->actor.fgName, 0, var_v0);
+        }
+    }
+}
 
 // TODO: this function includes an unreferenced float. It was forced to be in this function to
 // be able to build and match the file. It needs to be figured out where to put this float and
 // how to handle this when this file gets decompiled.
-#pragma GLOBAL_ASM("asm/jp/nonmatchings/overlays/actors/ovl_Ball/ac_ball/func_80969DE8_jp.s")
 
-#pragma GLOBAL_ASM("asm/jp/nonmatchings/overlays/actors/ovl_Ball/ac_ball/func_80969FBC_jp.s")
+// RO_FLT_8096A96C_jp 0.200000003
+
+s16 angl_add_table[2] = {
+    0x100,
+    0x400,
+};
+
+void aBALL_set_spd_relations_in_water(Ball* this, Game_Play* game_play) {
+    UNUSED UNK_TYPE1 pad[0x4];
+    xyz_t flowPos;
+    s16 yRot;
+    f32 waterHeight;
+    s32 yDiffAngle;
+
+    waterHeight = mCoBG_GetWaterHeight_File(this->actor.world.pos, "../ac_ball.c", 0x3B5);
```

Comment:
```suggestion
    waterHeight = mCoBG_GetWaterHeight_File(this->actor.world.pos, "../ac_ball.c", 949);
```
These kinds of functions pass the file name and the line number as an argument, we usually convert the line number to decimal.